Ledger.java

```diff
@@ -0,0 +1,268 @@
+package doubleEntry;
+
+import doubleEntryComponents.Agent;
+import doubleEntryComponents.Bank;
+import doubleEntryComponents.actions.Action;
+import doubleEntryComponents.contracts.Asset;
+import doubleEntryComponents.contracts.Contract;
+import doubleEntryComponents.contracts.Loan;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+
+/**
+ * This is the main class implementing double entry accounting. All public operations provided by this class
+ * are given as a double entry operation, i.e. a pair of (dr, cr) operations.
+ *
+ * A ledger contains a set of accounts, and is the interface between an agent and its accounts. Agents cannot
+ * directly interact with accounts other than via a ledger.
+ *
+ * A simple economic agent will usually have a single Ledger, whereas complex firms and banks can have several.
+ */
+public class Ledger {
+
+    public Ledger(Bank bank) {
+        accounts = new HashSet<>(); // A list of accounts
+
+        // Subsets of the list for quicker searching
+        assets = new HashSet<>();
+        liabilities = new HashSet<>();
+        equity = new HashSet<>();
+
+        // A hashmap relating types of contracts to the account they should be sitting in
+        defaultAssetAccounts = new HashMap<>();
+        // Note that separate hashmaps must exist for assets and liabilities since some contracts can be either
+        defaultLiabilityAccounts = new HashMap<>();
+        this.bank = bank;
+    }
+
+
+    private Bank bank;
+    private HashSet<Account> accounts;
+    private HashSet<Account> assets;
+    private HashSet<Account> liabilities;
+    private HashSet<Account> equity;
+    private HashMap<Class<? extends doubleEntryComponents.contracts.Contract>,Account> defaultAssetAccounts;
+    private HashMap<Class<? extends doubleEntryComponents.contracts.Contract>,Account> defaultLiabilityAccounts;
+    private Account defaultCashAccount;
+    private Account defaultEquityAccount;
+
+    public double getAssetValue() {
+        double assetTotal = 0;
+        for (Account assetAccount : assets) {
+            assetTotal+=assetAccount.getTotal();
+        }
+        return assetTotal;
+    }
+
+
+    public double getLiabilityValue() {
```

Comment:
Same question as with getAssetValue. Are contracts being added to the evaluation of an account?